<a href="https://colab.research.google.com/github/eppingera/EDGeS/blob/master/EDGe_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#EDGe Pipeline
This pipeline trains a Machine Learning classifier to determine if an inputed district maps is gerrymandered. It is trained on 2016 House of Representatives election results and district demographic data. Here, we use 80% of the data as a testing set. After this model is trained, a drawn district map can be inputed (ie a list of districts with the demographic features). Then, the model will classify which districts are "gerrymandered."

#Import Data

In [0]:
#all the numbers written on census
import io
import pandas as pd #data frame
import numpy as np #handle numebrs

#to make this notebook's output stable across runs garantees this shuffle will be same as next
np.random.seed(42)
url = 'https://raw.githubusercontent.com/eppingera/EDGeS/master/data.csv'

#import data into pandas dataframe
data = pd.read_csv(url)

data source from:

https://hub.arcgis.com/datasets/ff48bbae433442a38f6c635b8c7baf72_0/data

https://www.politico.com/2016-election/results/map/house/

In [0]:
#View the first 5 entries
data.head()

NUMBER   ID                  NAME  ...  GenX_1  GenBoom  GenSilent
0       1  101  Alabama - District 1  ...  125902   174921      70716
1       2  102  Alabama - District 2  ...  124496   163829      66245
2       3  103  Alabama - District 3  ...  126777   169330      64252
3       4  104  Alabama - District 4  ...  126242   173988      74147
4       5  105  Alabama - District 5  ...  133395   173788      68177

[5 rows x 94 columns]

In [0]:
#get data type and number of instances for all the atributes
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 436 entries, 0 to 435
Data columns (total 94 columns):
NUMBER             436 non-null int64
ID                 436 non-null int64
NAME               436 non-null object
NUM_REP_VOTES      386 non-null float64
NUM_DEM_VOTES      394 non-null float64
NUM_OTH_VOTES      186 non-null float64
TOTAL_NUM_VOTES    436 non-null int64
ST_ABBREV          436 non-null object
TOTPOP_CY          436 non-null int64
HHPOP_CY           436 non-null int64
FAMPOP_CY          436 non-null int64
GQPOP_CY           436 non-null int64
POPDENS_CY         436 non-null float64
TOTHH_CY           436 non-null int64
AVGHHSZ_CY         436 non-null float64
FAMHH_CY           436 non-null int64
AVGFMSZ_CY         436 non-null float64
TOTHU_CY           436 non-null int64
OWNER_CY           436 non-null int64
RENTER_CY          436 non-null int64
VACANT_CY          436 non-null int64
POPGRW10CY         436 non-null float64
HHGRW10CY          436 non-null float64
FAM

**Note:**
1.   the NUM_REP_VOTES, NUM_DEM_VOTES, & NUM_OTH_VOTES categories are not full. That's not a problem because we are going to use these categories to create new columns that we will actually use. 
2.   The NAME and ST_ABBREV categories are objects. This is not a problem because we are going to drop them in the ML model anyway. 
3.   Otherwise, all categoires have 436 data numberical points.
4.   Also, there are only 435 voting members of the House of Representativies. This data set contains the non-voting DC representative.
5.   Complete list of attributes is in the first data source link above -- other than the NUM_REP_VOTES, NUM_DEM_VOTES, NUM_OTH_VOTES, TOT_NUM_VOTES (number of Republican, Democrate, other, and total votes for each district)

To summarize the features of the dataset, they are various metrics of: 


*   Age demographics and distribution
*   Socioeconomic statuses and distribution
*   Percent homeownership, renting, vacancy
*   Population density
*   Population growth rate



#Remove Uncontested Seats

Remove any row where the House of Representatives' seat was uncontested that year (aka we have not voting results, so we can't use the data point).

In [0]:
#remove_uncon() removes any row where the seat was not contested in 2016
#returns a new pandas data fram without uncontested districts

def remove_uncon():
  uncon = []
  for i in range(len(data)):
    if (data.iloc[i,6] == 0):
      uncon.append(i)
  return data.drop(uncon)

#data_2 now holds the new data frame without uncontested districts
data_2 = remove_uncon()

#just checking :)
data_2.head(15)
      

NUMBER   ID                           NAME  ...  GenX_1  GenBoom  GenSilent
1        2  102           Alabama - District 2  ...  124496   163829      66245
2        3  103           Alabama - District 3  ...  126777   169330      64252
4        5  105           Alabama - District 5  ...  133395   173788      68177
5        6  106           Alabama - District 6  ...  135548   164428      63197
7        8  200  Alaska Congressional District  ...  132460   171412      38932
8        9  401           Arizona - District 1  ...  119793   179623      69419
9       10  402           Arizona - District 2  ...  117122   183936      92072
11      12  404           Arizona - District 4  ...  118202   214742     112655
12      13  405           Arizona - District 5  ...  145805   163373      77557
13      14  406           Arizona - District 6  ...  140344   195710      78118
14      15  407           Arizona - District 7  ...  129190   113131      31612
15      16  408           Arizona - District 8  ...  132657   174489     103425
16      17  409           Arizona - District 9  ...  130847   151117      52984
17      18  501          Arkansas - District 1  ...  128071   182369      81344
18      19  502          Arkansas - District 2  ...  136987   175282      68268

[15 rows x 94 columns]

#Remove NaN Values

One challenge with this data set is the NaN is the NUM Votes columns. We will use a simple method to remove these values and make them 0 instead.

In [0]:
data_2.fillna(0,inplace=True) #fill NaN values with 0
data_2.head(15)

NUMBER   ID                           NAME  ...  GenX_1  GenBoom  GenSilent
1        2  102           Alabama - District 2  ...  124496   163829      66245
2        3  103           Alabama - District 3  ...  126777   169330      64252
4        5  105           Alabama - District 5  ...  133395   173788      68177
5        6  106           Alabama - District 6  ...  135548   164428      63197
7        8  200  Alaska Congressional District  ...  132460   171412      38932
8        9  401           Arizona - District 1  ...  119793   179623      69419
9       10  402           Arizona - District 2  ...  117122   183936      92072
11      12  404           Arizona - District 4  ...  118202   214742     112655
12      13  405           Arizona - District 5  ...  145805   163373      77557
13      14  406           Arizona - District 6  ...  140344   195710      78118
14      15  407           Arizona - District 7  ...  129190   113131      31612
15      16  408           Arizona - District 8  ...  132657   174489     103425
16      17  409           Arizona - District 9  ...  130847   151117      52984
17      18  501          Arkansas - District 1  ...  128071   182369      81344
18      19  502          Arkansas - District 2  ...  136987   175282      68268

[15 rows x 94 columns]

#Generate Labels

In [0]:
#rep_win() returns an int array 
#(0 if dem's win, 1 if rep's win, others won no seats)
#this will be added as a column to the pd data frame
def rep_win():
  returner = []
  for i in range(len(data_2)):
    if (data_2.iloc[i,3] > data_2.iloc[i,4]):
      returner.append(1)
    else:
      returner.append(0)
  return returner

In [0]:
#add REP_WIN as a Column to the pd datafram after TOTAL_NUM_VOTES
data_2.insert(7, "REP_WIN", rep_win(), True) 

In [0]:
#check to see if that worked....
data_2.head(20)

NUMBER   ID                           NAME  ...  GenX_1  GenBoom  GenSilent
1        2  102           Alabama - District 2  ...  124496   163829      66245
2        3  103           Alabama - District 3  ...  126777   169330      64252
4        5  105           Alabama - District 5  ...  133395   173788      68177
5        6  106           Alabama - District 6  ...  135548   164428      63197
7        8  200  Alaska Congressional District  ...  132460   171412      38932
8        9  401           Arizona - District 1  ...  119793   179623      69419
9       10  402           Arizona - District 2  ...  117122   183936      92072
11      12  404           Arizona - District 4  ...  118202   214742     112655
12      13  405           Arizona - District 5  ...  145805   163373      77557
13      14  406           Arizona - District 6  ...  140344   195710      78118
14      15  407           Arizona - District 7  ...  129190   113131      31612
15      16  408           Arizona - District 8  ...  132657   174489     103425
16      17  409           Arizona - District 9  ...  130847   151117      52984
17      18  501          Arkansas - District 1  ...  128071   182369      81344
18      19  502          Arkansas - District 2  ...  136987   175282      68268
19      20  503          Arkansas - District 3  ...  136733   165754      67804
20      21  504          Arkansas - District 4  ...  124061   185249      82856
21      22  601        California - District 1  ...  113168   199100      85577
22      23  602        California - District 2  ...  127841   204102      76916
23      24  603        California - District 3  ...  123424   158081      57367

[20 rows x 95 columns]

In [0]:
#by_state() returns 3 lists
# 1) state abbreviations
# 2) total votes in each state
# 3) republican votes in each state
# 4) total seats in each state
# 5) republican seats in each state
def by_state():
  #list that will ultimately be returned
  state_abbrev = []
  total_votes = []
  rep_votes = []
  total_seats = []
  rep_seats = []

  #the 0th elements of the state array hols the first state abbrevatiation (AL)
  state_abbrev.append(data_2.iloc[0,8])
  total_votes.append(data_2.iloc[0,6])
  rep_votes.append(data_2.iloc[0,3])
  total_seats.append(1)
  rep_seats.append(data_2.iloc[0,7])

  #counter will be used to iterate through the returned arrays
  counter = 0

  for i in range(len(data_2)):
    #if the distric we are at is in the same state
    if (data_2.iloc[i,8] == state_abbrev[counter]):
      #add its values to the counter's position in the arrays below
      total_votes[counter] = total_votes[counter] + data_2.iloc[i,6]
      rep_votes[counter] = rep_votes[counter] + data_2.iloc[i,3]
      total_seats[counter] = total_seats[counter] + 1
      rep_seats[counter] = rep_seats[counter] + data_2.iloc[i,7]

    #if its a new state
    else:
      #add one to the counter
      counter = counter + 1
      #append its values to the ends of the returner arrays
      state_abbrev.append(data_2.iloc[i,8])
      total_votes.append(data_2.iloc[i,6])
      rep_votes.append(data_2.iloc[i,3])
      total_seats.append(1)
      rep_seats.append(data_2.iloc[i,7])
      #if there is another district in this state, this values will be increase accordingly
  
  return {
      "state_abbrev" : state_abbrev,
      "total_votes" : total_votes, 
      "rep_votes" : rep_votes,
      "total_seats" : total_seats, 
      "rep_seats" : rep_seats
  }    

In [0]:
#create a dictionary to hold the state_abbrev, total_votes, rep_votes, total_seats, rep_seats arrays
state_info = by_state()

In [0]:
data_2.head()

NUMBER   ID                           NAME  ...  GenX_1  GenBoom  GenSilent
1       2  102           Alabama - District 2  ...  124496   163829      66245
2       3  103           Alabama - District 3  ...  126777   169330      64252
4       5  105           Alabama - District 5  ...  133395   173788      68177
5       6  106           Alabama - District 6  ...  135548   164428      63197
7       8  200  Alaska Congressional District  ...  132460   171412      38932

[5 rows x 95 columns]

#HUGE: Setting the gerrymandering threshold
Here we can set the threshold for gerrymandering. Our metric for gerrymanding is the difference percent of seats won by a party and the percent of the total or overall vote that party won. The threshold is the difference in these percentages that is considered gerrymandering.

In [0]:
#let's set the threshold to 15% for now
threshold = 0.15

data_2.head()

NUMBER   ID                           NAME  ...  GenX_1  GenBoom  GenSilent
1       2  102           Alabama - District 2  ...  124496   163829      66245
2       3  103           Alabama - District 3  ...  126777   169330      64252
4       5  105           Alabama - District 5  ...  133395   173788      68177
5       6  106           Alabama - District 6  ...  135548   164428      63197
7       8  200  Alaska Congressional District  ...  132460   171412      38932

[5 rows x 95 columns]

In [0]:
#is_gerrymandered() returns whether or not a distric is gerrymandered
#it's based on the threshold set above
#this list will serve as the labeles for the ML model

def is_gerrymandered():
  is_gerrymandered = []
  for i in range(len(data_2)):
    #loc holds the index in the dictionary's arrays that we will access
    loc = (state_info['state_abbrev']).index(data_2.iloc[i,8])

    #percent_seats holds the #rep_seats / #total seats
    percent_seats = (state_info['rep_seats'])[loc] / (state_info['total_seats'])[loc]

    #percent_votes holds the #rep_votes / #total votes
    percent_votes = (state_info['rep_votes'])[loc] / (state_info['total_votes'])[loc]

    #if the difference in the %seats and %votes is greater than threshold, label 1
    if (abs(percent_seats - percent_votes) > threshold):
      is_gerrymandered.append(1)

    else:
      is_gerrymandered.append(0)

  return is_gerrymandered

In [0]:
labels = is_gerrymandered()
data_2.insert(3, "Labels", labels, True) 

#Quick aside

In [0]:
#For fun, let's see which states are the most gerrymandered :)
#negative: more gerrymandered for democrats
#postive: more gerrymandered for republicans

class State:
  abbrev = ''
  diff = 0

  #constructor
  def __init__(self, abbrev, diff):
    self.abbrev = abbrev
    self.diff = diff

  #toString method
  def __str__(self):
    return self.abbrev + ", " + str(self.diff)

def most():
  states = []
  #for each state, determine the difference in percent of votes and seats won
  for i in range(len((state_info['state_abbrev']))):
    percent_seats = (state_info['rep_seats'])[i] / (state_info['total_seats'])[i]
    percent_votes = (state_info['rep_votes'])[i] / (state_info['total_votes'])[i]

    abbrev = (state_info['state_abbrev'])[i]
    diff = percent_seats - percent_votes

    state = State(abbrev, diff)
    states.append(state)

  #for every object, compute lambda function (diff field)
  #sorts based on the diff attribute
  states.sort(key=lambda x: (x.diff))

  return states

for s in most():
  print(s)


NH, -0.44097509099435056
DE, -0.4096359188098708
CT, -0.36269804470608596
MA, -0.2518551199512158
MD, -0.2234346665446515
NV, -0.21221281909763456
HI, -0.20716408097856373
OR, -0.1878798960649744
NM, -0.10656724150083918
MN, -0.09247514969032128
CA, -0.08999454549518876
IL, -0.08043361392800258
NJ, -0.04821919298546046
WA, -0.046305255115762056
NY, -0.023720440695768397
VT, 0.0
ME, 0.018223473310051197
AZ, 0.06512458136515864
FL, 0.07931391699287915
CO, 0.08339266730186834
TX, 0.12520390143432447
LA, 0.12805988304038363
GA, 0.14513230266705845
KY, 0.15008436891663346
TN, 0.15288910574777292
MO, 0.1592253831531799
MI, 0.16207557664819877
VA, 0.17035186525661616
MS, 0.17102123315172646
OH, 0.17291763947421734
WI, 0.2084185737769978
IA, 0.21231180435742614
PA, 0.2170467858691194
NC, 0.2364900024996225
IN, 0.25617534131621733
SC, 0.2722437745558771
AR, 0.2878391182806851
ND, 0.30764207370467667
OK, 0.3255734769465888
ID, 0.34434523458942434
WV, 0.3514858019458711
AL, 0.3556644973210574
SD,

#Machine Learning Models

In [0]:
data_2.head()

NUMBER   ID                           NAME  ...  GenX_1  GenBoom  GenSilent
1       2  102           Alabama - District 2  ...  124496   163829      66245
2       3  103           Alabama - District 3  ...  126777   169330      64252
4       5  105           Alabama - District 5  ...  133395   173788      68177
5       6  106           Alabama - District 6  ...  135548   164428      63197
7       8  200  Alaska Congressional District  ...  132460   171412      38932

[5 rows x 96 columns]

In [0]:
#Drop what we won't use in our ML model
data_2 = data_2.drop(["NUMBER", "ID", "NAME", "NUM_REP_VOTES", "NUM_DEM_VOTES", "NUM_OTH_VOTES", "ST_ABBREV"], axis=1) 

In [0]:
data_2.head()

Labels  TOTAL_NUM_VOTES  REP_WIN  ...  GenX_1  GenBoom  GenSilent
1       1           246090        1  ...  124496   163829      66245
2       1           284291        1  ...  126777   169330      64252
4       1           306368        1  ...  133395   173788      68177
5       1           322193        1  ...  135548   164428      63197
7       1           248775        1  ...  132460   171412      38932

[5 rows x 89 columns]

In [0]:
#Create a test set of data to work with.
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(data_2, test_size=0.2, random_state=42)

#create labels and features for both the training and testing sets
x_train = train_set.drop("Labels", axis = 1)
y_train = train_set["Labels"]

x_test = test_set.drop("Labels", axis = 1)
y_test = test_set["Labels"]

##SGD Classifier

In [0]:
#SGDClassifier (Stochastic Gradient Descent):

#Import the SGDClassifier
from sklearn.linear_model import SGDClassifier

#Import the cross_val_score
from sklearn.model_selection import cross_val_predict

#Define the classifier with max_iter = 5, tol = -np.infty, random_state = 42
sgd_clf = SGDClassifier(max_iter = 5, tol = -np.infty, random_state = 42)

#Fit the sgd classifer
sgd_clf.fit(x_train, y_train)

#hold the sgd predictions
sgd_train_pred = cross_val_predict(sgd_clf, x_train, y_train, cv = 10)

In [0]:
#SGD F1 Score
from sklearn.metrics import f1_score

f1_score(y_train, sgd_train_pred)

0.5529411764705883

In [0]:
#SGD ROC AUC
from sklearn.metrics import roc_auc_score

roc_auc_score(y_train, sgd_train_pred)

0.5406060606060606

##SVC Classifier

In [0]:
#SVCClassifier (Support Vector Classification)
from sklearn.svm import SVC

svm_clf = SVC(gamma="auto")
svm_clf.fit(x_train, y_train)
svm_train_pred = cross_val_predict(svm_clf, x_train, y_train, cv=10)

In [0]:
#SVC F1 Score
f1_score(y_train, svm_train_pred)

0.0

In [0]:
#SVC ROC AUC Score
roc_auc_score(y_train, svm_train_pred)

0.5

##Random Forest Classifier

In [0]:
#Random Forest:

#Import the RandomForestClassifier:
from sklearn.ensemble import RandomForestClassifier

#Define the classifier with n_estimators = 100 and random_state = 42 and
forest_clf = RandomForestClassifier(n_estimators = 100, random_state = 42)

#Fit the classifier to X_train and y_train
forest_clf.fit(x_train, y_train)

#Get scores for the random forest classifier, cv = 10 and scoring 
forest_train_pred = cross_val_predict(forest_clf, x_train, y_train, cv=10)

In [0]:
#Forest F1 Score
f1_score(y_train, forest_train_pred)

0.7919463087248321

In [0]:
#Fores ROC AUC
roc_auc_score(y_train, forest_train_pred)

0.8081060606060606

#Test Random Forest Model
The Randcom Forest Classifier appears to be best for this dataset. Now, we can test the classifier on the testing data and measure its accuracy.

In [0]:
#test Random Forest classifier on the test data set
y_test_pred = forest_clf.predict(x_test)

In [0]:
#F1 score for test data set
f1_score(y_test, y_test_pred)

0.8648648648648649

In [0]:
#ROC AUC Score for test data
roc_auc_score(y_test, y_test_pred)

0.8768768768768769

#End Result:

According to this finding.... https://www.sciencedirect.com/science/article/pii/S1556086415306043

The Random Forest ROC AUC Value is 0.88, which considered "excellent" for an ML model.